File: frontend/src/types/scanDelta.ts
export type DeltaCategory = "findings" | "components" | "crypto";
export type DeltaChange =
  | "added"
  | "removed"
  | "version_changed"
  | "license_changed";

export interface ScanDeltaTotals {
  added: number;
  removed: number;
  unchanged: number;
  changed: number;
  by_severity: Record<string, number>;
  by_type: Record<string, number>;
}

export interface FindingDeltaItem {
  change: "added" | "removed";
  finding_id: string;
  finding_type: string;
  severity: string;
  title: string;
  component: string | null;
  cve_id: string | null;
  file_path: string | null;
  first_seen: string | null;
}

export interface ComponentDeltaItem {
  change: "added" | "removed" | "version_changed" | "license_changed";
  name: string;
  purl: string | null;
  version: string | null;
  from_version: string | null;
  to_version: string | null;
  license: string | null;
  from_license: string | null;
  to_license: string | null;
}

export interface CryptoDeltaItem {
  change: "added" | "removed";
  name: string;
  variant: string | null;
  primitive: string | null;
  locations: string[];
  asset_count: number;
}

export type DeltaItem = FindingDeltaItem | ComponentDeltaItem | CryptoDeltaItem;

export interface ScanDeltaReachability {
  coverable_count: number;
  analyzed_count: number;
}

// Which build a side of the comparison actually is. A symbolic side ("release", "head") resolves
// server-side, and a release side follows the rescan chain, so the requested id need not be it.
export interface ScanDeltaSide {
  scan_id: string;
  branch: string | null;
  commit_hash: string | null;
  created_at: string | null;
}

export interface ScanDeltaResponse {
  from_scan_id: string;
  to_scan_id: string;
  from_side?: ScanDeltaSide | null;
  to_side?: ScanDeltaSide | null;
  project_id: string;
  category: DeltaCategory;
  totals: ScanDeltaTotals;
  page: number;
  page_size: number;
  total_pages: number;
  items: DeltaItem[];
  // Null means the scan reports no reachability at all, which is distinct from zero coverage.
  from_reachability?: ScanDeltaReachability | null;
  to_reachability?: ScanDeltaReachability | null;
  // Findings a waiver hides in whole or in part on each side.
  from_waived_excluded: number;
  to_waived_excluded: number;
  // Added and removed items the comparison would not have produced had no waiver applied.
  waiver_only_changes: number;
}

File: backend/app/schemas/scan_delta.py
"""Response envelopes for the unified scan-delta API (findings, components, crypto)."""

from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


class DeltaCategory(str, Enum):
    """Top-level category of delta the response describes."""

    FINDINGS = "findings"
    COMPONENTS = "components"
    CRYPTO = "crypto"


class DeltaChange(str, Enum):
    """Kinds of change an individual delta item can represent."""

    ADDED = "added"
    REMOVED = "removed"
    VERSION_CHANGED = "version_changed"
    LICENSE_CHANGED = "license_changed"


class ScanDeltaTotals(BaseModel):
    """Aggregate counts for a scan-delta response."""

    model_config = ConfigDict(extra="forbid")

    added: int = 0
    removed: int = 0
    unchanged: int = 0
    # Only set for the components category.
    changed: int = 0
    # Findings-only breakdowns.
    by_severity: dict[str, int] = Field(default_factory=dict)
    by_type: dict[str, int] = Field(default_factory=dict)


class FindingDeltaItem(BaseModel):
    """A single added/removed finding between two scans."""

    model_config = ConfigDict(extra="forbid")

    change: Literal["added", "removed"]
    finding_id: str
    finding_type: str
    severity: str
    title: str
    component: str | None = None
    cve_id: str | None = None
    file_path: str | None = None
    first_seen: datetime | None = None


class ComponentDeltaItem(BaseModel):
    """A single component change between two scans."""

    model_config = ConfigDict(extra="forbid")

    change: Literal["added", "removed", "version_changed", "license_changed"]
    name: str
    purl: str | None = None
    version: str | None = None
    from_version: str | None = None
    to_version: str | None = None
    license: str | None = None
    from_license: str | None = None
    to_license: str | None = None


class CryptoDeltaItem(BaseModel):
    """A single crypto asset change between two scans."""

    model_config = ConfigDict(extra="forbid")

    change: Literal["added", "removed"]
    name: str
    variant: str | None = None
    primitive: str | None = None
    locations: list[str] = Field(default_factory=list)
    asset_count: int = 1


DeltaItem = FindingDeltaItem | ComponentDeltaItem | CryptoDeltaItem


class ScanDeltaResponse(BaseModel):
    """Unified response envelope for the scan-delta endpoint."""

    model_config = ConfigDict(extra="forbid")

    from_scan_id: str
    to_scan_id: str
    project_id: str
    category: DeltaCategory
    totals: ScanDeltaTotals
    page: int = 1
    page_size: int = 50
    total_pages: int = 1
    items: list[DeltaItem] = Field(default_factory=list)
